text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+        Tcp(Address),
+    }
+    ```
+
+    Note, that future updates may support different protocols, which is why this enum is declared as `non_exhaustive`, see `Unresolved Questions` section.
+
+    ```rust
+    // Infallible construction of a `Url` from an `Address` and a `Protocol`.
+    fn new(addr: Address, proto: Protocol) -> Self { ... }
+
+    // Fallible construction of a `Url` from `&str`.
+    async fn from_url_str(url: &str) -> Result<Self> { ... }
+    ```
+
+    Note, that the second function is `async`. That is because of possible delays when performing domain name resolution.
+
+
+
+## `Endpoint` and `EndpointId`
+
+To model the remote part of a connection, this crate introduces the `Endpoint` type:
+
+```rust
+struct Endpoint {
+    // The id of the endpoint.
+    id: EndpointId,
+
+    // The address of the endpoint.
+    address: Address,
+
+    // The protocol used to communicate with that endpoint.
+    protocol: Protocol,
+}
+```
+
+Note: In a peer-to-peer (p2p) network peers usually have two endpoints initially as they are actively trying to connect, but also need to accept connections from their peers. This crate is agnostic about how to handle duplicate connections as this is usually resolved by a handshaking protocol defined at a higher layer.
+
+To uniquely identify an `Endpoint`, this crate proposes the `EndpointId` type, which can be implemented as a wrapper around an `Address`.
+
+```rust
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
+struct EndpointId(Address);
+```
+
+Note, that this `EndpointId` should be `Clone` and `Copy`, and requires to implement `Eq`, `PartialEq`, and `Hash`, because it is used as a hashmap key in several instances.
+
+## `Command`
+
+`Command`s are messages that are supposed to be issued by higher level business logic like the protocol layer. They are implemented as an `enum`, which in Rust is one way of expressing a polymorphic type:
+
+```rust
+enum Command {
+    // Adds an endpoint to the list.
+    AddEndpoint { url: Url, ... },
+
+    // Removes an endpoint from the list.
+    RemoveEndpoint { epid: EndpointId, ... },
+
+    // Attempts to connect to an endpoint.
+    Connect { epid: EndpointId, ... },
+
+    // Disconnects from an endpoint.
+    Disconnect { epid: EndpointId, ... },
+
+    // Sends a message to a particular endpoint.
+    SendMessage { epid: EndpointId, bytes: Vec<u8>, ... },
+
+    // Sends a message to multiple endpoints.
+    MulticastMessage { epids: Vec<EndpointId>, bytes: Vec<u8>, ...},
+
+    // Sends a message to all endpoints in its list.
+    BroadcastMessage { bytes: Vec<u8>, ... },
+}
+```
+
+This `enum` makes the things the consumer can do with the crate very explicit and descriptive, and also easily extensible.
+
+### `Event`
+
+Similarly to `Command`s, the different kinds of `Event`s in the system are implemented as an `enum` allowing various types of concrete events being sent over event channels:
+
+```rust
+enum Event {
```

Comment:
I think I would made this one non_exhaustive since a consumer is expected to matcher over it.

Reply:
ah, very good point and foresight actually. :+1:

Reply:
done